packages/editor-ui/src/composables/useExecutionDebugging.ts

```diff
@@ -0,0 +1,139 @@
+import { computed } from 'vue';
+import { useRouter } from 'vue-router';
+import { useI18n, useMessage, useToast } from '@/composables';
+import {
+	DEBUG_PAYWALL_MODAL_KEY,
+	EnterpriseEditionFeature,
+	MODAL_CONFIRM,
+	VIEWS,
+} from '@/constants';
+import { useSettingsStore, useUIStore, useWorkflowsStore } from '@/stores';
+import type { INodeUi } from '@/Interface';
+
+export const useExecutionDebugging = () => {
+	const router = useRouter();
+	const i18n = useI18n();
+	const message = useMessage();
+	const toast = useToast();
+	const workflowsStore = useWorkflowsStore();
+	const settingsStore = useSettingsStore();
+	const uiStore = useUIStore();
+
+	const isDebugEnabled = computed(() =>
+		settingsStore.isEnterpriseFeatureEnabled(EnterpriseEditionFeature.DebugInEditor),
+	);
+
+	const applyExecutionData = async (executionId: string): Promise<void> => {
+		const execution = await workflowsStore.getExecution(executionId);
+		const workflow = workflowsStore.getCurrentWorkflow();
+		const workflowNodes = workflowsStore.getNodes();
+
+		if (!execution?.data?.resultData) {
+			return;
+		}
+
+		const { runData } = execution.data.resultData;
+
+		const executionNodeNames = Object.keys(runData);
+		const missingNodeNames = executionNodeNames.filter(
+			(name) => !workflowNodes.some((node) => node.name === name),
+		);
+
+		// Using the pinned data of the workflow to check if the node is pinned
+		// because workflowsStore.getCurrentWorkflow() returns a cached workflow without the updated pinned data
+		const workflowPinnedNodeNames = Object.keys(workflowsStore.workflow.pinData ?? {});
+		const matchingPinnedNodeNames = executionNodeNames.filter((name) =>
+			workflowPinnedNodeNames.includes(name),
+		);
+		const matchingPinnedNodeNamesToHtmlList = `<ul class="ml-l">${matchingPinnedNodeNames
+			.map((name) => `<li>${name}</li>`)
+			.join('')}</ul>`;
+
+		if (matchingPinnedNodeNames.length > 0) {
+			const overWritePinnedDataConfirm = await message.confirm(
+				i18n.baseText('nodeView.confirmMessage.debug.message', {
+					interpolate: { nodeNames: matchingPinnedNodeNamesToHtmlList },
+				}),
+				i18n.baseText('nodeView.confirmMessage.debug.headline'),
+				{
+					type: 'warning',
+					confirmButtonText: i18n.baseText('nodeView.confirmMessage.debug.confirmButtonText'),
+					cancelButtonText: i18n.baseText('nodeView.confirmMessage.debug.cancelButtonText'),
+					dangerouslyUseHTMLString: true,
+					customClass: 'matching-pinned-nodes-confirmation',
+				},
+			);
+
+			if (overWritePinnedDataConfirm === MODAL_CONFIRM) {
+				matchingPinnedNodeNames.forEach((name) => {
+					const node = workflowsStore.getNodeByName(name);
+					if (node) {
+						workflowsStore.unpinData({ node });
+					}
+				});
+			} else {
+				await router.push({
+					name: VIEWS.EXECUTION_PREVIEW,
+					params: { name: workflow.id, executionId },
+				});
+				return;
+			}
+		}
```

Comment:
The HTML string here leaves us open to XSS attacks via node names. We should render the HTML via render function (supported by element-plus).

Reply:
You're right, couldn't leave it like that ;)